src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,820 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// This DemiBuffer type is designed to be a common abstraction defining the behavior of data buffers in Demikernel.
+// It currently supports two underlying types of buffers: heap-allocated and DPDK-allocated.  The basic operations on
+// DemiBuffers are designed to have equivalent behavior (effects on the data), regardless of the underlying buffer type.
+// In particular, len(), adjust(), trim(), clone(), and split() are present and designed to behave the same regardless.
+//
+// The constructors/destructors, however, are necessarily different.  For DPDK-allocated buffers, a MBuf is expected
+// to be allocated externally and provided to the DemiBuffer's "from_mbuf" constructor.  A MBuf can also be extracted
+// from a DPDK allocated DemiBuffer via the "into_mbuf" routine.
+//
+// Note: if compiled without the "libdpdk" feature defined, the DPDK-specific functionality won't be present.
+
+// Note on buffer chain support:
+// DPDK has a concept of MBuf chaining where multiple MBufs may be linked together to form a "packet".  While the
+// DemiBuffer routines for heap-allocated buffers also now support this functionality, it isn't yet exposed via the
+// DemiBuffer interface.
+// ToDo: Expose buffer chain support once we have a solid use case.
+
+// Note on intrusive queueing:
+// Since all DemiBuffer types keep the metadata for each "view" in a separate allocated region, they can be queued
+// using intrusive links (i.e. have a link field in the metadata).
+// ToDo: Expose calls to get/set a linking field.
+
+// Note on the allocation functions:
+// This code currently uses std::alloc() and std::dealloc() to allocate/free things from the heap.  Note that the Rust
+// documentation says that these functions are expected to be deprecated in favor of their respective methods of the
+// "Global" type when it and the "Allocator" trait become stable.
+use crate::runtime::fail::Fail;
+#[cfg(feature = "libdpdk")]
+use ::dpdk_rs::{
+    rte_mbuf,
+    rte_mempool,
+    rte_pktmbuf_adj,
+    rte_pktmbuf_clone,
+    rte_pktmbuf_free,
+    rte_pktmbuf_trim,
+};
+#[cfg(feature = "libdpdk")]
+use ::std::mem;
+use ::std::{
+    alloc::{
+        alloc,
+        dealloc,
+        handle_alloc_error,
+        Layout,
+    },
+    marker::PhantomData,
+    mem::size_of,
+    num::NonZeroUsize,
+    ops::{
+        Deref,
+        DerefMut,
+    },
+    ptr::{
+        self,
+        null_mut,
+        NonNull,
+    },
+    slice,
+};
+
+// Cache line size, defined here for code clarity.
+// While this is arguably architecture dependent, and thus should be pulled from some architecture-specific information,
+// this value (64 bytes) is used by all architectures we're likely to care about (and is also assumed by DPDK).
+const CACHE_LINE_SIZE: usize = 64;
```

Comment:
Could you move this to `pal::arch`? This is an architecture constant. Preferably, rely on the `crate` for this as well.

Reply:
Do you mean create a pal::arch?  I only see a pal::linux in the current dev branch.

Reply:
@BrianZill yes, create another module `pal::arch` and place architectural dependent code there ;-)

Reply:
Okay, I added an arch.rs under pal and put the CPU_DATA_CACHE_LINE_SIZE constant in it.